src/ast/SemanticChecksVisitor.java

```diff
@@ -290,115 +309,215 @@ else if(ownerExprType.equals("ast.IdentifierExpr")){
+        whileStatement.body().accept(this);
+        // we don't know if the condition so we assume it didn't change anything in upper scope
+        definitelyInitialized.pop();
     }
 
     @Override
     public void visit(SysoutStatement sysoutStatement) {
-
+        sysoutStatement.arg().accept(this);
     }
 
     @Override
     public void visit(AssignStatement assignStatement) {
-
+        SymbolTable declTable = STLookup.findDeclTable(assignStatement.lv(), forest, assignStatement.enclosingScope(), programST);
+        VariableIntroduction assigneeDecl = (VariableIntroduction) STLookup.getDeclNode(declTable, assignStatement.lv());
+        AstType assigneeType = assigneeDecl.type();
+        AstType assignedValueType = getExprType(assignStatement.rv());
+        // TODO continue: req 15
+        assignStatement.rv().accept(this);
+        definitelyInitialized.peek().add(assignStatement.lv());
     }
 
     @Override
     public void visit(AssignArrayStatement assignArrayStatement) {
+        if (!definitelyInitialized.peek().contains(assignArrayStatement.lv())) {
+            visitResult = false;
+            return;
+        }
 
+        assignArrayStatement.index().accept(this);
+        assignArrayStatement.rv().accept(this);
+    }
+
+    private void visit(BinaryExpr e) {
+        e.e1().accept(this);
+        e.e2().accept(this);
     }
 
     @Override
     public void visit(AndExpr e) {
-
+        visit((BinaryExpr)e);
     }
 
     @Override
     public void visit(LtExpr e) {
-
+        visit((BinaryExpr)e);
     }
 
     @Override
     public void visit(AddExpr e) {
-
+        visit((BinaryExpr)e);
     }
 
     @Override
     public void visit(SubtractExpr e) {
-
+        visit((BinaryExpr)e);
     }
 
     @Override
     public void visit(MultExpr e) {
-
+        visit((BinaryExpr)e);
     }
 
     @Override
     public void visit(ArrayAccessExpr e) {
-
+        e.arrayExpr().accept(this);
+        e.indexExpr().accept(this);
     }
 
     @Override
     public void visit(ArrayLengthExpr e) {
-
+        // req 12 - x.length where x is an int[]
+        if (!(getExprType(e.arrayExpr()) instanceof IntArrayAstType)) {
+            visitResult = false;
+            return;
+        }
+        e.arrayExpr().accept(this);
     }
 
     @Override
     public void visit(MethodCallExpr e) {
+        // req 11 - a method call is invoked on a ref type (this, new, or
+        // var / formal /  field that are ref types
+        AstType ownerType = getExprType(e.ownerExpr());
+        // req 9 - must be a reference type
+        if (!(ownerType instanceof RefType)) {
+            visitResult = false;
+            return;
+        }
+
+        // req 10 method signature check
+        String className = ((RefType) ownerType).id();
+        SymbolTable classTable = programST.getSymbol(className, false).enclosedScope();
+        if (!classTable.contains(e.methodId(), true)) {
```

Comment:
not good enough - could be inherited in which case it is legal but will not show up in SymbolTable.
try vtable instead
I'm thinking maybe we should have a map instead of vtable (keys will be names and values will be MethodDecl)